tools/src/abstract_server/local_index.rs

```diff
@@ -0,0 +1,123 @@
+use async_trait::async_trait;
+use flate2::read::GzDecoder;
+use serde_json::{from_str, Value};
+use std::io::Read;
+use tokio::fs::File;
+use tokio::io::AsyncReadExt;
+
+use super::server_interface::{AbstractServer, ErrorDetails, ErrorLayer, Result, ServerError};
+
+use crate::config::{load, TreeConfigPaths};
+
+/// IO errors amount to a 404 for our purposes which means a sticky problem.
+impl From<std::io::Error> for ServerError {
+    fn from(err: std::io::Error) -> ServerError {
+        ServerError::StickyProblem(ErrorDetails {
+            layer: ErrorLayer::ServerLayer,
+            message: err.to_string(),
+        })
+    }
+}
+
+/// Read newline-delimited JSON that's been gzip-compressed.
+async fn read_gzipped_ndjson_from_file(path: &str) -> Result<Vec<Value>> {
+    let mut f = File::open(path).await?;
+    // We read the entirety to a buffer because
+    // https://github.com/serde-rs/json/issues/160 suggests that the buffered
+    // reader performance is likely to be much worse.
+    //
+    // Alternately, the flate2 library can be built with the "tokio" feature and
+    // we could do fully buffered decoding.
+    let mut buffer = Vec::new();
+    f.read_to_end(&mut buffer).await?;
+
+    let mut gz = GzDecoder::new(&buffer[..]);
+
+    let mut raw_str = String::new();
+    gz.read_to_string(&mut raw_str)?;
+
+    // let mut raw_str = String::new();
+    // f.read_to_string(&mut raw_str).await?;
+
+    Ok(raw_str.lines().filter_map(|s| from_str(s).ok()).collect())
```

Comment:
Okay, I addressed this here and in remote_server by changing:
```rust
Ok(raw_str.lines().filter_map(|s| from_str(s).ok()).collect())
```
into (after rustfmt)
```rust
    raw_str
        .lines()
        .map(|s| from_str(s).map_err(|e| ServerError::from(e)))
        .collect()
```
after also implementing `impl From<serde_json::Error> for ServerError` in `server_interface.rs` since serde is a common error between all the implementations.

At first I just tried without the `map_err`, but this resulted in an error because `FromIterator` didn't know how to transform the error when going from `Vec<Result<T, E>>` to `Result<Vec<T>, E>` because the E's weren't the same.  I'm not sure if I actually could have implemented a custom FromIterator, but it seemed like a lot of typing, and in general I understand Vec::collect() to generally require help with types, so this seemed reasonably straightforward.

I'm definitely open to hints on more idiomatic options here!